x/sync/README.md

```diff
@@ -1,3 +1,57 @@
+
+1. `SyncGetRangeProofRequest`
+2. `RangeProof`
+3. `SyncGetChangeProofRequest`
+4. `SyncGetChangeProofResponse`
+
+These message types are defined in `avalanchego/proto/sync.proto`.
+For more information on range proofs and change proofs, see their definitions in `avalanchego/merkledb/proof.go`.
+
+### `SyncGetRangeProofRequest`
+
+This message is sent from the client to the server to request a range proof for a given key range and root ID. That is, the client says, "Give me the key-value pairs that were in this key range when the database had this root."
+
+### `RangeProof`
+
+This message is sent from the server to the client in response to a `SyncGetRangeProofRequest`. It contains the key-value pairs that were in the requested key range when the database had the requested root, as well as a proof that the key-value pairs are correct.
+
+### `SyncGetChangeProofRequest`
+
+This message is sent from the client to the server to request a change proof between the given root IDs. That is, the client says, "Give me the key-value pairs that changed between the time the database had this root and that root."
+
+### `SyncGetChangeProofResponse`
+
+This message is sent from the server to the client in response to a `SyncGetChangeProofRequest`. If the server had sufficient history to generate a change proof, it contains a change proof that contains the key-value pairs that changed between the requested roots. If the server did not have sufficient history to generate a change proof, it contains a range proof that contains the key-value pairs that were in the database when the database had the latter root.
+
+## Algorithm
+
+When the client starts with an empty database, it has no key-value pairs. It requests from a server a range proof for the entire database. The server replies with a range proof, which the client verifies. If it's valid, the key-value pairs in the proof are written to the database. If it's not, the client drops the proof and requests the proof from another server. 
+
+A range proof sent by a server may be valid but not contain all of the key-value pairs in the requested range. For example, a client might request all the key-value pairs in [`start`, `end`] but only receive those in range [`start`, `end'`] where `end'` < `end`. There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response. Unless the database is very small, this means that the range proof the client receives in response to its range proof request for the entire database will not contain all of the key-value pairs in the database.
+
+For each key range, the sync client keeps track of the root ID of the last time it synced that key range. For example, it will store information that says something like, "I have all of the key-value pairs that were in range [`start`, `end`] when the database's root was `rootID`" for some keys `start` and `end`, and some database `rootID`.
```

Comment:
I think this could use some clarity rewrites.  It technically only tracks it for each received proof rather than all key ranges.  It might also be worth specifying that it is the remote DB's Root ID that is recorded rather than whatever the local one is at the moment.

Reply:
>  It technically only tracks it for each received proof rather than all key ranges

Isn't this semantically the same, though?

> It might also be worth specifying that it is the remote DB's Root ID that is recorded rather than whatever the local one is at the moment. 

done

Reply:
You could have data in your local db from before syncing started, so the known roots are only for received proofs during this sync cycle rather than from all time.